{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*NEED TO WORK ON* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading raw data for HH2022 ...\n",
      "Raw data loaded.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Evidence IDs</th>\n",
       "      <th>Phosphosite IDs</th>\n",
       "      <th>Leading Proteins</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "      <th>Modifications</th>\n",
       "      <th>Modified sequence</th>\n",
       "      <th>Oxidation (M) Probabilities</th>\n",
       "      <th>Phospho (STY) Probabilities</th>\n",
       "      <th>...</th>\n",
       "      <th>Reporter intensity 2</th>\n",
       "      <th>Reporter intensity 3</th>\n",
       "      <th>Reporter intensity 4</th>\n",
       "      <th>Reporter intensity 5</th>\n",
       "      <th>Reporter intensity 6</th>\n",
       "      <th>Reporter intensity 7</th>\n",
       "      <th>Reporter intensity 8</th>\n",
       "      <th>Reporter intensity 9</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>13593;13594;13595;19565</td>\n",
       "      <td>Q9NQS7</td>\n",
       "      <td>INCENP</td>\n",
       "      <td>AAAAAAAATMALAAPSSPTPESPTMLTK</td>\n",
       "      <td>28</td>\n",
       "      <td>Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_AAAAAAAATMALAAPS(ph)SPTPESPTM(ox)LTK_</td>\n",
       "      <td>AAAAAAAATM(0.022)ALAAPSSPTPESPTM(0.978)LTK</td>\n",
       "      <td>AAAAAAAAT(0.003)MALAAPS(0.303)S(0.303)PT(0.303...</td>\n",
       "      <td>...</td>\n",
       "      <td>16835.0</td>\n",
       "      <td>16028.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1347.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13593;13594;13595;19565</td>\n",
       "      <td>Q9NQS7</td>\n",
       "      <td>INCENP</td>\n",
       "      <td>AAAAAAAATMALAAPSSPTPESPTMLTK</td>\n",
       "      <td>28</td>\n",
       "      <td>Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_AAAAAAAATMALAAPS(ph)SPTPESPTM(ox)LTK_</td>\n",
       "      <td>AAAAAAAATM(0.049)ALAAPSSPTPESPTM(0.951)LTK</td>\n",
       "      <td>AAAAAAAAT(0.003)MALAAPS(0.304)S(0.304)PT(0.304...</td>\n",
       "      <td>...</td>\n",
       "      <td>11121.0</td>\n",
       "      <td>10992.0</td>\n",
       "      <td>648.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1057.60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3338.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>13593;13594;13595;19565</td>\n",
       "      <td>Q9NQS7</td>\n",
       "      <td>INCENP</td>\n",
       "      <td>AAAAAAAATMALAAPSSPTPESPTMLTK</td>\n",
       "      <td>28</td>\n",
       "      <td>Phospho (STY)</td>\n",
       "      <td>_AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAAAAAAAT(0.008)MALAAPS(0.425)S(0.425)PT(0.124...</td>\n",
       "      <td>...</td>\n",
       "      <td>11061.0</td>\n",
       "      <td>13143.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1116.30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1156.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>13593;13594;13595;19565</td>\n",
       "      <td>Q9NQS7</td>\n",
       "      <td>INCENP</td>\n",
       "      <td>AAAAAAAATMALAAPSSPTPESPTMLTK</td>\n",
       "      <td>28</td>\n",
       "      <td>Phospho (STY)</td>\n",
       "      <td>_AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAAAAAAAT(0.003)MALAAPS(0.288)S(0.288)PT(0.288...</td>\n",
       "      <td>...</td>\n",
       "      <td>10076.0</td>\n",
       "      <td>10433.0</td>\n",
       "      <td>3426.60</td>\n",
       "      <td>507.41</td>\n",
       "      <td>1465.00</td>\n",
       "      <td>2967.9</td>\n",
       "      <td>2213.90</td>\n",
       "      <td>2013.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>13593;13594;13595;19565</td>\n",
       "      <td>Q9NQS7</td>\n",
       "      <td>INCENP</td>\n",
       "      <td>AAAAAAAATMALAAPSSPTPESPTMLTK</td>\n",
       "      <td>28</td>\n",
       "      <td>2 Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_AAAAAAAATM(ox)ALAAPS(ph)SPTPESPTM(ox)LTK_</td>\n",
       "      <td>AAAAAAAATM(1)ALAAPSSPTPESPTM(1)LTK</td>\n",
       "      <td>AAAAAAAATMALAAPS(0.325)S(0.325)PT(0.325)PES(0....</td>\n",
       "      <td>...</td>\n",
       "      <td>15124.0</td>\n",
       "      <td>15420.0</td>\n",
       "      <td>253.76</td>\n",
       "      <td>2105.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>1087.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165472</th>\n",
       "      <td>165472</td>\n",
       "      <td>13863</td>\n",
       "      <td>Q9NW97</td>\n",
       "      <td>TMEM51</td>\n",
       "      <td>YYVPSYEEVMNTNYSEAR</td>\n",
       "      <td>18</td>\n",
       "      <td>Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_YYVPS(ph)YEEVM(ox)NTNYSEAR_</td>\n",
       "      <td>YYVPSYEEVM(1)NTNYSEAR</td>\n",
       "      <td>Y(0.006)Y(0.006)VPS(0.931)Y(0.057)EEVMNT(0.001...</td>\n",
       "      <td>...</td>\n",
       "      <td>18551.0</td>\n",
       "      <td>20730.0</td>\n",
       "      <td>15898.00</td>\n",
       "      <td>3334.50</td>\n",
       "      <td>7259.30</td>\n",
       "      <td>11527.0</td>\n",
       "      <td>8817.70</td>\n",
       "      <td>4230.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165473</th>\n",
       "      <td>165473</td>\n",
       "      <td>13863</td>\n",
       "      <td>Q9NW97</td>\n",
       "      <td>TMEM51</td>\n",
       "      <td>YYVPSYEEVMNTNYSEAR</td>\n",
       "      <td>18</td>\n",
       "      <td>Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_YYVPS(ph)YEEVM(ox)NTNYSEAR_</td>\n",
       "      <td>YYVPSYEEVM(1)NTNYSEAR</td>\n",
       "      <td>Y(0.001)Y(0.001)VPS(0.952)Y(0.047)EEVMNTNYSEAR</td>\n",
       "      <td>...</td>\n",
       "      <td>19759.0</td>\n",
       "      <td>20859.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1910.30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>954.29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165474</th>\n",
       "      <td>165474</td>\n",
       "      <td>13863</td>\n",
       "      <td>Q9NW97</td>\n",
       "      <td>TMEM51</td>\n",
       "      <td>YYVPSYEEVMNTNYSEAR</td>\n",
       "      <td>18</td>\n",
       "      <td>Oxidation (M),Phospho (STY)</td>\n",
       "      <td>_YYVPS(ph)YEEVM(ox)NTNYSEAR_</td>\n",
       "      <td>YYVPSYEEVM(1)NTNYSEAR</td>\n",
       "      <td>Y(0.008)Y(0.008)VPS(0.839)Y(0.145)EEVMNTNYSEAR</td>\n",
       "      <td>...</td>\n",
       "      <td>23503.0</td>\n",
       "      <td>25199.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>339.41</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1082.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165475</th>\n",
       "      <td>165475</td>\n",
       "      <td>13863</td>\n",
       "      <td>Q9NW97</td>\n",
       "      <td>TMEM51</td>\n",
       "      <td>YYVPSYEEVMNTNYSEAR</td>\n",
       "      <td>18</td>\n",
       "      <td>Phospho (STY)</td>\n",
       "      <td>_YYVPS(ph)YEEVMNTNYSEAR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y(0.003)Y(0.003)VPS(0.822)Y(0.172)EEVMNTNYSEAR</td>\n",
       "      <td>...</td>\n",
       "      <td>12115.0</td>\n",
       "      <td>12867.0</td>\n",
       "      <td>2475.60</td>\n",
       "      <td>1366.80</td>\n",
       "      <td>879.97</td>\n",
       "      <td>2017.5</td>\n",
       "      <td>2357.70</td>\n",
       "      <td>1418.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165476</th>\n",
       "      <td>165476</td>\n",
       "      <td>13863</td>\n",
       "      <td>Q9NW97</td>\n",
       "      <td>TMEM51</td>\n",
       "      <td>YYVPSYEEVMNTNYSEAR</td>\n",
       "      <td>18</td>\n",
       "      <td>Phospho (STY)</td>\n",
       "      <td>_YYVPS(ph)YEEVMNTNYSEAR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YYVPS(0.956)Y(0.043)EEVMNTNYSEAR</td>\n",
       "      <td>...</td>\n",
       "      <td>15901.0</td>\n",
       "      <td>13102.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>716.49</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165477 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Evidence IDs          Phosphosite IDs Leading Proteins Gene Names  \\\n",
       "0                  0  13593;13594;13595;19565           Q9NQS7     INCENP   \n",
       "1                  1  13593;13594;13595;19565           Q9NQS7     INCENP   \n",
       "2                  2  13593;13594;13595;19565           Q9NQS7     INCENP   \n",
       "3                  3  13593;13594;13595;19565           Q9NQS7     INCENP   \n",
       "4                  4  13593;13594;13595;19565           Q9NQS7     INCENP   \n",
       "...              ...                      ...              ...        ...   \n",
       "165472        165472                    13863           Q9NW97     TMEM51   \n",
       "165473        165473                    13863           Q9NW97     TMEM51   \n",
       "165474        165474                    13863           Q9NW97     TMEM51   \n",
       "165475        165475                    13863           Q9NW97     TMEM51   \n",
       "165476        165476                    13863           Q9NW97     TMEM51   \n",
       "\n",
       "                            Sequence  Length                  Modifications  \\\n",
       "0       AAAAAAAATMALAAPSSPTPESPTMLTK      28    Oxidation (M),Phospho (STY)   \n",
       "1       AAAAAAAATMALAAPSSPTPESPTMLTK      28    Oxidation (M),Phospho (STY)   \n",
       "2       AAAAAAAATMALAAPSSPTPESPTMLTK      28                  Phospho (STY)   \n",
       "3       AAAAAAAATMALAAPSSPTPESPTMLTK      28                  Phospho (STY)   \n",
       "4       AAAAAAAATMALAAPSSPTPESPTMLTK      28  2 Oxidation (M),Phospho (STY)   \n",
       "...                              ...     ...                            ...   \n",
       "165472            YYVPSYEEVMNTNYSEAR      18    Oxidation (M),Phospho (STY)   \n",
       "165473            YYVPSYEEVMNTNYSEAR      18    Oxidation (M),Phospho (STY)   \n",
       "165474            YYVPSYEEVMNTNYSEAR      18    Oxidation (M),Phospho (STY)   \n",
       "165475            YYVPSYEEVMNTNYSEAR      18                  Phospho (STY)   \n",
       "165476            YYVPSYEEVMNTNYSEAR      18                  Phospho (STY)   \n",
       "\n",
       "                                 Modified sequence  \\\n",
       "0           _AAAAAAAATMALAAPS(ph)SPTPESPTM(ox)LTK_   \n",
       "1           _AAAAAAAATMALAAPS(ph)SPTPESPTM(ox)LTK_   \n",
       "2               _AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_   \n",
       "3               _AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_   \n",
       "4       _AAAAAAAATM(ox)ALAAPS(ph)SPTPESPTM(ox)LTK_   \n",
       "...                                            ...   \n",
       "165472                _YYVPS(ph)YEEVM(ox)NTNYSEAR_   \n",
       "165473                _YYVPS(ph)YEEVM(ox)NTNYSEAR_   \n",
       "165474                _YYVPS(ph)YEEVM(ox)NTNYSEAR_   \n",
       "165475                    _YYVPS(ph)YEEVMNTNYSEAR_   \n",
       "165476                    _YYVPS(ph)YEEVMNTNYSEAR_   \n",
       "\n",
       "                       Oxidation (M) Probabilities  \\\n",
       "0       AAAAAAAATM(0.022)ALAAPSSPTPESPTM(0.978)LTK   \n",
       "1       AAAAAAAATM(0.049)ALAAPSSPTPESPTM(0.951)LTK   \n",
       "2                                              NaN   \n",
       "3                                              NaN   \n",
       "4               AAAAAAAATM(1)ALAAPSSPTPESPTM(1)LTK   \n",
       "...                                            ...   \n",
       "165472                       YYVPSYEEVM(1)NTNYSEAR   \n",
       "165473                       YYVPSYEEVM(1)NTNYSEAR   \n",
       "165474                       YYVPSYEEVM(1)NTNYSEAR   \n",
       "165475                                         NaN   \n",
       "165476                                         NaN   \n",
       "\n",
       "                              Phospho (STY) Probabilities  ...  \\\n",
       "0       AAAAAAAAT(0.003)MALAAPS(0.303)S(0.303)PT(0.303...  ...   \n",
       "1       AAAAAAAAT(0.003)MALAAPS(0.304)S(0.304)PT(0.304...  ...   \n",
       "2       AAAAAAAAT(0.008)MALAAPS(0.425)S(0.425)PT(0.124...  ...   \n",
       "3       AAAAAAAAT(0.003)MALAAPS(0.288)S(0.288)PT(0.288...  ...   \n",
       "4       AAAAAAAATMALAAPS(0.325)S(0.325)PT(0.325)PES(0....  ...   \n",
       "...                                                   ...  ...   \n",
       "165472  Y(0.006)Y(0.006)VPS(0.931)Y(0.057)EEVMNT(0.001...  ...   \n",
       "165473     Y(0.001)Y(0.001)VPS(0.952)Y(0.047)EEVMNTNYSEAR  ...   \n",
       "165474     Y(0.008)Y(0.008)VPS(0.839)Y(0.145)EEVMNTNYSEAR  ...   \n",
       "165475     Y(0.003)Y(0.003)VPS(0.822)Y(0.172)EEVMNTNYSEAR  ...   \n",
       "165476                   YYVPS(0.956)Y(0.043)EEVMNTNYSEAR  ...   \n",
       "\n",
       "        Reporter intensity 2  Reporter intensity 3  Reporter intensity 4  \\\n",
       "0                    16835.0               16028.0                  0.00   \n",
       "1                    11121.0               10992.0                648.59   \n",
       "2                    11061.0               13143.0                  0.00   \n",
       "3                    10076.0               10433.0               3426.60   \n",
       "4                    15124.0               15420.0                253.76   \n",
       "...                      ...                   ...                   ...   \n",
       "165472               18551.0               20730.0              15898.00   \n",
       "165473               19759.0               20859.0                  0.00   \n",
       "165474               23503.0               25199.0                  0.00   \n",
       "165475               12115.0               12867.0               2475.60   \n",
       "165476               15901.0               13102.0                  0.00   \n",
       "\n",
       "        Reporter intensity 5  Reporter intensity 6  Reporter intensity 7  \\\n",
       "0                       0.00                  0.00                   0.0   \n",
       "1                       0.00               1057.60                   0.0   \n",
       "2                       0.00               1116.30                   0.0   \n",
       "3                     507.41               1465.00                2967.9   \n",
       "4                    2105.20                  0.00                1054.0   \n",
       "...                      ...                   ...                   ...   \n",
       "165472               3334.50               7259.30               11527.0   \n",
       "165473               1910.30                  0.00                   0.0   \n",
       "165474                339.41                  0.00                   0.0   \n",
       "165475               1366.80                879.97                2017.5   \n",
       "165476                716.49                  0.00                   0.0   \n",
       "\n",
       "        Reporter intensity 8  Reporter intensity 9  Unnamed: 25   　  \n",
       "0                       0.00                1347.1          NaN NaN  \n",
       "1                       0.00                3338.3          NaN NaN  \n",
       "2                       0.00                1156.9          NaN NaN  \n",
       "3                    2213.90                2013.1          NaN NaN  \n",
       "4                    1087.00                   0.0          NaN NaN  \n",
       "...                      ...                   ...          ...  ..  \n",
       "165472               8817.70                4230.9          NaN NaN  \n",
       "165473                954.29                   0.0          NaN NaN  \n",
       "165474                  0.00                1082.8          NaN NaN  \n",
       "165475               2357.70                1418.2          NaN NaN  \n",
       "165476                  0.00                   0.0          NaN NaN  \n",
       "\n",
       "[165477 rows x 27 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'HH2022'\n",
    "\n",
    "print('Loading raw data for', dataset, '...')\n",
    "data = pd.read_csv('/Users/maryamkoddus/Documents/maryam-ko-QMUL-MSc-Project/01_input_data/raw_data/41598_2022_8430_MOESM12_ESM.csv', header=0)\n",
    "print('Raw data loaded.')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene Names  Modifications                   Modified sequence\n",
      "2          INCENP  Phospho (STY)  _AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_\n",
      "3          INCENP  Phospho (STY)  _AAAAAAAATMALAAPS(ph)SPTPESPTMLTK_\n",
      "9           SAMD1  Phospho (STY)          _AAAAAATAPPS(ph)PGPAQPGPR_\n",
      "10          SAMD1  Phospho (STY)          _AAAAAAT(ph)APPSPGPAQPGPR_\n",
      "11          SAMD1  Phospho (STY)          _AAAAAATAPPS(ph)PGPAQPGPR_\n",
      "...           ...            ...                                 ...\n",
      "165468   ARHGEF35  Phospho (STY)      _YYS(ph)PCEEHPAETNQNEGAESGTIR_\n",
      "165469    ARHGEF5  Phospho (STY)      _YYSPCEEHPAET(ph)NQNEGSESGTIR_\n",
      "165470    ARHGEF5  Phospho (STY)      _YYSPCEEHPAET(ph)NQNEGSESGTIR_\n",
      "165475     TMEM51  Phospho (STY)            _YYVPS(ph)YEEVMNTNYSEAR_\n",
      "165476     TMEM51  Phospho (STY)            _YYVPS(ph)YEEVMNTNYSEAR_\n",
      "\n",
      "[131402 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "filtered_data = data[data['Modifications'].str.contains(r'^Phospho \\(STY\\)$', case=False, na=False)]\n",
    "data = filtered_data[filtered_data['Modifications'].str.count(',') == 0]\n",
    "\n",
    "print(data[['Gene Names', 'Modifications', 'Modified sequence']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene Names               Modified sequence  \\\n",
      "9           SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "12          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "13          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "14          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "15          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "...           ...                             ...   \n",
      "165467   ARHGEF35  _YYS(ph)PCEEHPAETNQNEGAESGTIR_   \n",
      "165468   ARHGEF35  _YYS(ph)PCEEHPAETNQNEGAESGTIR_   \n",
      "165469    ARHGEF5  _YYSPCEEHPAET(ph)NQNEGSESGTIR_   \n",
      "165470    ARHGEF5  _YYSPCEEHPAET(ph)NQNEGSESGTIR_   \n",
      "165476     TMEM51        _YYVPS(ph)YEEVMNTNYSEAR_   \n",
      "\n",
      "                              Phospho (STY) Probabilities  \n",
      "9                      AAAAAAT(0.011)APPS(0.989)PGPAQPGPR  \n",
      "12                     AAAAAAT(0.005)APPS(0.995)PGPAQPGPR  \n",
      "13                     AAAAAAT(0.001)APPS(0.999)PGPAQPGPR  \n",
      "14                     AAAAAAT(0.001)APPS(0.999)PGPAQPGPR  \n",
      "15                     AAAAAAT(0.011)APPS(0.989)PGPAQPGPR  \n",
      "...                                                   ...  \n",
      "165467  Y(0.001)Y(0.001)S(0.997)PCEEHPAET(0.002)NQNEGA...  \n",
      "165468  Y(0.001)Y(0.001)S(0.957)PCEEHPAET(0.042)NQNEGA...  \n",
      "165469             YYS(0.061)PCEEHPAET(0.938)NQNEGSESGTIR  \n",
      "165470               YYS(0.03)PCEEHPAET(0.97)NQNEGSESGTIR  \n",
      "165476                   YYVPS(0.956)Y(0.043)EEVMNTNYSEAR  \n",
      "\n",
      "[50410 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "def has_high_phospho_probability(prob_str):\n",
    "    # Extract all numbers within parentheses, \n",
    "    probabilities = re.findall(r'\\((\\d+\\.\\d+)\\)', prob_str)\n",
    "    \n",
    "    # Convert strings to float and check if any value is greater than 0.85\n",
    "    return any(float(prob) > 0.85 for prob in probabilities)\n",
    "\n",
    "data = filtered_data[filtered_data['Phospho (STY) Probabilities'].apply(has_high_phospho_probability)]\n",
    "\n",
    "print(data[['Gene Names', 'Modified sequence', 'Phospho (STY) Probabilities']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene Names               Modified sequence  \\\n",
      "9           SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "12          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "13          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "14          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "15          SAMD1      _AAAAAATAPPS(ph)PGPAQPGPR_   \n",
      "...           ...                             ...   \n",
      "165467   ARHGEF35  _YYS(ph)PCEEHPAETNQNEGAESGTIR_   \n",
      "165468   ARHGEF35  _YYS(ph)PCEEHPAETNQNEGAESGTIR_   \n",
      "165469    ARHGEF5  _YYSPCEEHPAET(ph)NQNEGSESGTIR_   \n",
      "165470    ARHGEF5  _YYSPCEEHPAET(ph)NQNEGSESGTIR_   \n",
      "165476     TMEM51        _YYVPS(ph)YEEVMNTNYSEAR_   \n",
      "\n",
      "                              Phospho (STY) Probabilities Amino Acid  \n",
      "9                      AAAAAAT(0.011)APPS(0.989)PGPAQPGPR        [S]  \n",
      "12                     AAAAAAT(0.005)APPS(0.995)PGPAQPGPR        [S]  \n",
      "13                     AAAAAAT(0.001)APPS(0.999)PGPAQPGPR        [S]  \n",
      "14                     AAAAAAT(0.001)APPS(0.999)PGPAQPGPR        [S]  \n",
      "15                     AAAAAAT(0.011)APPS(0.989)PGPAQPGPR        [S]  \n",
      "...                                                   ...        ...  \n",
      "165467  Y(0.001)Y(0.001)S(0.997)PCEEHPAET(0.002)NQNEGA...        [S]  \n",
      "165468  Y(0.001)Y(0.001)S(0.957)PCEEHPAET(0.042)NQNEGA...        [S]  \n",
      "165469             YYS(0.061)PCEEHPAET(0.938)NQNEGSESGTIR        [T]  \n",
      "165470               YYS(0.03)PCEEHPAET(0.97)NQNEGSESGTIR        [T]  \n",
      "165476                   YYVPS(0.956)Y(0.043)EEVMNTNYSEAR        [S]  \n",
      "\n",
      "[50410 rows x 4 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1v/y_z33m9d25d2tszlfrvnp2300000gn/T/ipykernel_12697/3352352718.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data['Amino Acid'] = data['Modified sequence'].apply(extract_modified_amino_acid)\n"
     ]
    }
   ],
   "source": [
    "def extract_modified_amino_acid(modified_seq):\n",
    "    modified_amino_acids = re.findall(r'([A-Z])\\(ph\\)', modified_seq)  # Only match phosphorylation\n",
    "    return [aa for aa in modified_amino_acids]\n",
    "\n",
    "\n",
    "data['Amino Acid'] = data['Modified sequence'].apply(extract_modified_amino_acid)\n",
    "\n",
    "print(data[['Gene Names', 'Modified sequence', 'Phospho (STY) Probabilities', 'Amino Acid']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene Names                  Sequence  StartPosition\n",
      "9           SAMD1      AAAAAATAPPSPGPAQPGPR          150.0\n",
      "12          SAMD1      AAAAAATAPPSPGPAQPGPR          150.0\n",
      "13          SAMD1      AAAAAATAPPSPGPAQPGPR          150.0\n",
      "14          SAMD1      AAAAAATAPPSPGPAQPGPR          150.0\n",
      "15          SAMD1      AAAAAATAPPSPGPAQPGPR          150.0\n",
      "...           ...                       ...            ...\n",
      "165467   ARHGEF35  YYSPCEEHPAETNQNEGAESGTIR          181.0\n",
      "165468   ARHGEF35  YYSPCEEHPAETNQNEGAESGTIR          181.0\n",
      "165469    ARHGEF5  YYSPCEEHPAETNQNEGSESGTIR          181.0\n",
      "165470    ARHGEF5  YYSPCEEHPAETNQNEGSESGTIR          181.0\n",
      "165476     TMEM51        YYVPSYEEVMNTNYSEAR          128.0\n",
      "\n",
      "[50410 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1v/y_z33m9d25d2tszlfrvnp2300000gn/T/ipykernel_12697/2766028544.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dataset['StartPosition'] = dataset['Gene Names'].map(positions_dict)\n"
     ]
    }
   ],
   "source": [
    "def find_position_in_gene(dataset, seq_column, fasta_path):\n",
    "    positions_dict = {}\n",
    "    \n",
    "    # Parse the FASTA file\n",
    "    fasta_sequence = list(SeqIO.parse(open(fasta_path), \"fasta\"))\n",
    "    \n",
    "    # Iterate over rows in the dataset\n",
    "    for i, row in dataset.iterrows():\n",
    "        gene_sequence = row[seq_column]  # Get the sequence from the dataset\n",
    "        \n",
    "        for seq_record in fasta_sequence:\n",
    "            # Check if the sequence is part of the gene in the FASTA record\n",
    "            start_position = str(seq_record.seq).find(gene_sequence)\n",
    "            \n",
    "            # If a match is found, store the position\n",
    "            if start_position != -1:\n",
    "                positions_dict[row['Gene Names']] = start_position\n",
    "                break  # Stop once the position is found for the first matching gene\n",
    "    \n",
    "    # Add the start positions to the dataset as a new column\n",
    "    dataset['StartPosition'] = dataset['Gene Names'].map(positions_dict)\n",
    "    return dataset\n",
    "\n",
    "fasta_path = '/Users/maryamkoddus/Documents/maryam-ko-QMUL-MSc-Project/01_input_data/raw_data/UP000005640_9606.fasta'\n",
    "data = find_position_in_gene(data, 'Sequence', fasta_path)\n",
    "\n",
    "# Print the data with start positions\n",
    "print(data[['Gene Names', 'Sequence', 'StartPosition']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene Names Amino Acid  StartPosition                  Sequence  \\\n",
      "9           SAMD1        [S]          150.0      AAAAAATAPPSPGPAQPGPR   \n",
      "12          SAMD1        [S]          150.0      AAAAAATAPPSPGPAQPGPR   \n",
      "13          SAMD1        [S]          150.0      AAAAAATAPPSPGPAQPGPR   \n",
      "14          SAMD1        [S]          150.0      AAAAAATAPPSPGPAQPGPR   \n",
      "15          SAMD1        [S]          150.0      AAAAAATAPPSPGPAQPGPR   \n",
      "...           ...        ...            ...                       ...   \n",
      "165467   ARHGEF35        [S]          181.0  YYSPCEEHPAETNQNEGAESGTIR   \n",
      "165468   ARHGEF35        [S]          181.0  YYSPCEEHPAETNQNEGAESGTIR   \n",
      "165469    ARHGEF5        [T]          181.0  YYSPCEEHPAETNQNEGSESGTIR   \n",
      "165470    ARHGEF5        [T]          181.0  YYSPCEEHPAETNQNEGSESGTIR   \n",
      "165476     TMEM51        [S]          128.0        YYVPSYEEVMNTNYSEAR   \n",
      "\n",
      "        Reporter intensity 0  Reporter intensity 1  Reporter intensity 2  \\\n",
      "9                    58392.0               53344.0               65403.0   \n",
      "12                   14635.0               13538.0                   0.0   \n",
      "13                  183920.0              200440.0              236420.0   \n",
      "14                   28570.0               24923.0               35335.0   \n",
      "15                  124420.0              126320.0              152890.0   \n",
      "...                      ...                   ...                   ...   \n",
      "165467               22109.0               18449.0               20685.0   \n",
      "165468                   NaN                   NaN                   NaN   \n",
      "165469                   0.0                7199.0                6900.4   \n",
      "165470                8164.7                5857.9                8512.2   \n",
      "165476               12949.0               10895.0               15901.0   \n",
      "\n",
      "        Reporter intensity 3  Reporter intensity 4  Reporter intensity 5  \\\n",
      "9                    71718.0                   0.0                449.38   \n",
      "12                   20225.0                1017.7                482.39   \n",
      "13                  227770.0               11130.0              18815.00   \n",
      "14                   34110.0                2442.2               1805.10   \n",
      "15                  172590.0                8819.3              12493.00   \n",
      "...                      ...                   ...                   ...   \n",
      "165467               17919.0               27444.0              13070.00   \n",
      "165468                   NaN                   NaN                   NaN   \n",
      "165469                9858.5               12174.0               8997.80   \n",
      "165470                8685.4               19816.0               9367.90   \n",
      "165476               13102.0                   0.0                716.49   \n",
      "\n",
      "        Reporter intensity 6  Reporter intensity 7  Reporter intensity 8  \\\n",
      "9                    12057.0                3290.1                   0.0   \n",
      "12                       0.0                   0.0                   0.0   \n",
      "13                   26802.0               12602.0                2677.3   \n",
      "14                    4455.6                5117.2                2528.5   \n",
      "15                   31432.0               10993.0                4322.1   \n",
      "...                      ...                   ...                   ...   \n",
      "165467               36607.0               27642.0               21322.0   \n",
      "165468                   NaN                   NaN                   NaN   \n",
      "165469                5765.2               15867.0               12735.0   \n",
      "165470                9901.8               22979.0               19071.0   \n",
      "165476                   0.0                   0.0                   0.0   \n",
      "\n",
      "        Reporter intensity 9  \n",
      "9                    47433.0  \n",
      "12                    7672.9  \n",
      "13                   77048.0  \n",
      "14                   23934.0  \n",
      "15                   51076.0  \n",
      "...                      ...  \n",
      "165467               18889.0  \n",
      "165468                   NaN  \n",
      "165469                3799.5  \n",
      "165470                5375.5  \n",
      "165476                   0.0  \n",
      "\n",
      "[50410 rows x 14 columns]\n"
     ]
    }
   ],
   "source": [
    "# Select the necessary columns: Gene Names, Modified Amino Acids, Sequence, and Reporter Intensity columns\n",
    "columns_to_display = ['Gene Names', 'Amino Acid', 'StartPosition', 'Sequence'] + [col for col in data.columns if 'Reporter intensity' in col]\n",
    "\n",
    "# Display the data with the selected columns\n",
    "print(data[columns_to_display])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Amino Acids'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/maryam-ko-QMUL-MSc-Project/.venv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3805\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3804\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m3805\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3806\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:167\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:196\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: 'Amino Acids'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[16]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Filtering out semi-colons from 'Amino acid', 'Positions within proteins', and 'Gene names' columns\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m data = data[~\u001b[43mdata\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mAmino Acids\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m.str.contains(\u001b[33m'\u001b[39m\u001b[33m;\u001b[39m\u001b[33m'\u001b[39m, na=\u001b[38;5;28;01mFalse\u001b[39;00m)]\n\u001b[32m      3\u001b[39m data = data[~data[\u001b[33m'\u001b[39m\u001b[33mGene Names\u001b[39m\u001b[33m'\u001b[39m].str.contains(\u001b[33m'\u001b[39m\u001b[33m;\u001b[39m\u001b[33m'\u001b[39m, na=\u001b[38;5;28;01mFalse\u001b[39;00m)]\n\u001b[32m      4\u001b[39m data\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/maryam-ko-QMUL-MSc-Project/.venv/lib/python3.13/site-packages/pandas/core/frame.py:4102\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4100\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.nlevels > \u001b[32m1\u001b[39m:\n\u001b[32m   4101\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._getitem_multilevel(key)\n\u001b[32m-> \u001b[39m\u001b[32m4102\u001b[39m indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4103\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[32m   4104\u001b[39m     indexer = [indexer]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/maryam-ko-QMUL-MSc-Project/.venv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3812\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3807\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   3808\u001b[39m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc.Iterable)\n\u001b[32m   3809\u001b[39m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[32m   3810\u001b[39m     ):\n\u001b[32m   3811\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[32m-> \u001b[39m\u001b[32m3812\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   3813\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[32m   3814\u001b[39m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[32m   3815\u001b[39m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[32m   3816\u001b[39m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[32m   3817\u001b[39m     \u001b[38;5;28mself\u001b[39m._check_indexing_error(key)\n",
      "\u001b[31mKeyError\u001b[39m: 'Amino Acids'"
     ]
    }
   ],
   "source": [
    "# Filtering out semi-colons from 'Amino acid', 'Positions within proteins', and 'Gene names' columns\n",
    "data = data[~data['Amino Acid'].str.contains(';', na=False)]\n",
    "data = data[~data['Gene Names'].str.contains(';', na=False)]\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter data\n",
    "data['Sequence'] = data['Sequence'].str.replace('_', '')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_seq_to_genename(dataset, seq_column):\n",
    "    '''\n",
    "    Maps amino acid sequences to gene names using the loaded fasta file.\n",
    "    \n",
    "    args:\n",
    "    =====\n",
    "    dataset: <pd.Dataframe> with a column of amino acid sequences\n",
    "    seq_column: <str> column name containing amino acid sequences\n",
    "    \n",
    "    out:\n",
    "    ====\n",
    "    dataset: <pd.Dataframe> with an additional column containing gene names\n",
    "    '''    \n",
    "\n",
    "    fasta_sequence = list(SeqIO.parse(open(f'/Users/maryamkoddus/Documents/maryam-ko-QMUL-MSc-Project/01_input_data/raw_data/UP000005640_9606.fasta'), \"fasta\"))\n",
    "    \n",
    "    \n",
    "    gene_dict = {}\n",
    "    \n",
    "    # iterate over rows in seq_column\n",
    "    for i in dataset[seq_column]:\n",
    "        print(i)\n",
    "        i_str = str(i)\n",
    "        for seq_record in fasta_sequence:\n",
    "            matches = re.findall(i_str, str(seq_record.seq))\n",
    "            if matches:\n",
    "                print(f\"Match found for sequence: {seq_record}\")\n",
    "                gene_name_match = re.search(r\"GN=(\\w+)\", seq_record.description)\n",
    "                print('Gene name match:', gene_name_match)\n",
    "                # gene_name_match = re.search(\"GN=(\\w+)\", seq_record.description)\n",
    "                if gene_name_match:\n",
    "                    gene_name = gene_name_match.group(1)\n",
    "                    gene_dict[i] = gene_name\n",
    "                    print(f\"Match found: {i_str} -> {gene_name}\")\n",
    "                else: \n",
    "                    print(f\"No gene name found in description for sequence: {i_str}\")\n",
    "    \n",
    "    # map sequences to gene names           \n",
    "    dataset['GeneName'] = dataset[seq_column].map(gene_dict) \n",
    "    print('Amino acid sequences matched to gene names.')\n",
    "    return dataset \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = match_seq_to_genename(data, 'Sequence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Phosphosite'] = data['Amino Acids'].astype(str) + '(' + data['StartPosition'].astype(str) + ')'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keepcols = ['Phosphosite'] + ['GeneName'] + [col for col in data.columns if 'Reporter intensity' in col]\n",
    "data = data[keepcols]\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Intensity_columns = [col for col in data.columns if 'Reporter intensity' in col]\n",
    "data[Intensity_columns] = data[Intensity_columns].apply(pd.to_numeric, errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log2_transform(dataset):\n",
    "    '''\n",
    "    Log2 transform a dataset.\n",
    "    \n",
    "    args:\n",
    "    =====\n",
    "    dataset: <pd.Dataframe>\n",
    "    \n",
    "    out:\n",
    "    ====\n",
    "    dataset: <pd.Dataframe> with log2 transformed values\n",
    "\n",
    "    '''\n",
    "    cols_to_transform = dataset[Intensity_columns]\n",
    "    dataset[Intensity_columns] = cols_to_transform.apply(np.log2)\n",
    "    print('Data has been log2 transformed.')\n",
    "    return dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = log2_transform(data)\n",
    "print(f\"DataFrame after log2 transformation:\\n{data}\") # Print the DataFrame after log2 transformation\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_phos_ID(dataset):\n",
    "    '''\n",
    "    Concatenates GeneName and Phosphosite columns.\n",
    "    \n",
    "    args:\n",
    "    =====\n",
    "    dataset: <pd.Dataframe> with columns 'GeneName' and 'Phosphosite'\n",
    "    \n",
    "    out:\n",
    "    ====\n",
    "    dataset: <pd.Dataframe> with 'phosphosite_ID' column and 'GeneName' + 'Phosphosite' columns dropped\n",
    "    '''\n",
    "    dataset.loc[:, 'phosphosite_ID'] = dataset['GeneName'].astype(str) + '_' + dataset['Phosphosite'].astype(str)\n",
    "    dataset = dataset.drop(columns=['Phosphosite', 'GeneName'])\n",
    "    print('Phosphosite IDs created.')\n",
    "    return dataset\n",
    "\n",
    "data = create_phos_ID(data) # call function to create phosphosite_ID column\n",
    "\n",
    "print('Phosphosite IDs created.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_phosID_col(data):\n",
    "    data = data[~data.phosphosite_ID.str.contains('nan', case=False, na=False)]\n",
    "    data = data[~data.phosphosite_ID.str.contains(';', case=False, na=False)]\n",
    "    data = data[~data.phosphosite_ID.str.contains('-', case=False, na=False)]\n",
    "\n",
    "    # Add this line to remove decimals from phosphosite_ID (e.g., S123.0 -> S123)\n",
    "    data['phosphosite_ID'] = data['phosphosite_ID'].apply(lambda x: re.sub(r'\\((\\d+)\\.0+\\)', r'(\\1)', x))\n",
    "    \n",
    "    data_grouped = data.groupby(by='phosphosite_ID')\n",
    "    \n",
    "    if len(data) != len(data_grouped):\n",
    "        numeric_cols = data.select_dtypes(include=[np.number]).columns.tolist()\n",
    "        non_numeric_cols = data.columns.difference(numeric_cols + ['phosphosite_ID']).tolist()\n",
    "        data_numeric = data_grouped[numeric_cols].mean()\n",
    "        data_categorical = data_grouped[non_numeric_cols].first().reset_index()\n",
    "        \n",
    "        # Merge numeric and non-numeric parts\n",
    "        data = pd.merge(data_categorical, data_numeric, on='phosphosite_ID')\n",
    "        print('Phosphosites with multiple measurements have been averaged')\n",
    "    else:\n",
    "        print('There are no phosphosites with multiple measurements')\n",
    "\n",
    "    # Replace inf values with NaNs\n",
    "    data = data.replace([np.inf, -np.inf], np.nan)\n",
    "    \n",
    "    # Ensure phosphosite_ID is first column\n",
    "    if data.columns[0] != 'phosphosite_ID':\n",
    "        phosphosite_ID = data.pop('phosphosite_ID')\n",
    "        data.insert(0, 'phosphosite_ID', phosphosite_ID)\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = clean_phosID_col(data)\n",
    "print(\"After cleaning phosphosite_ID column:\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv(f'/Users/maryamkoddus/Documents/maryam-ko-QMUL-MSc-Project/01_input_data/PreprocessedDatasets/HH2022.csv', index=False)\n",
    "\n",
    "\n",
    "print(dataset, 'has been saved to CSV successfully!', data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
